### DISCUSSION TITLE
Export @JsType to javascript env

### DISCUSSION THREAD

**Discussion by treblereel** [google/j2cl#213] [Q&A]
Hello, we are migrating from GWT2 to J2CL, and there is an issue with @JsType. In GWT2 it works well (ofc with generateJsInteropExports).

For now, I managed to export only methods, but I can't make class properties to be exported.

Let's say I have a simple POJO:

```
@JsType
public class Test {
    public String idddddddddddd = "OLOLO";
    public static String static_id = "qwerty";
    public static String staticTest() { return static_id; }
    public static void setStatic_id(String s) { static_id = s; }
    public final String getId() { return idddddddddddd; }
    public final void setId(String id) { this.idddddddddddd = id; }
}
```

My export is like this one:

```
goog.module('org.treblereel.Test$GWT3Export');

const EXPORT = goog.require('org.treblereel.Test');

goog.exportSymbol('org.treblereel.Test', EXPORT);
goog.exportSymbol('org.treblereel.Test.prototype.getId', EXPORT.prototype.getId);
goog.exportSymbol('org.treblereel.Test.prototype.setId', EXPORT.prototype.setId);
goog.exportSymbol('org.treblereel.Test.staticTest', EXPORT.staticTest);
goog.exportSymbol('org.treblereel.Test.setStatic_id', EXPORT.setStatic_id);
```

So, methods export works fine, but I need help managing the export fields. I tried `goog.export property like this:

```
goog.exportProperty(EXPORT.prototype, 'idddddddddddd', EXPORT.prototype.idddddddddddd);
goog.exportProperty(EXPORT, 'static_id', EXPORT.static_id);
```

But it doesn't work, it looks like I do something wrong.

**Comment by jiakuan** [comment]
I was also exploring how to export Java types including properties to JS environment. I found another way to keep the names (aka export). For example, if we put the following code in `xxx.native.js`, the function `onSubmit` would be accessible from browser window.

```
function onSubmit(token) {
......
}

window["onSubmit"] = onSubmit;
```

I'm not sure we could export Java classes and properties in a similar way, IF Closure API doesn't have good support for this.

**Reply by treblereel** [reply]
I have to preserve the package of the type (for instance, static prop must be available `org.treblereel.Test. static_id`), so I have to be very careful about goog.modules namespaces. So, assigning a type to the window object isn't an option for me ...

**Comment by gkdn** [comment]
Probably exporting static doesn't work since it gets evaluated during call of `goog.export` instead of linking to actual field. Maybe you can work around that with `Object.defineProperty` but I haven't tried that.

For general exporting guidance for exports and instance properties is here:
https://developers.google.com/closure/compiler/docs/externs-and-exports

**Reply by treblereel** [reply]
yeap, I can assign EXPORT.static_id to the object property, but in this case, I'll have two static props, that will be out of sync.

**Reply by gkdn** [reply]
The purpose of Object.defineProperty is to delegate to original definition so there wouldn't be extra state to go out of sync.
